cmd/parca-agent/main.go

```diff
@@ -102,71 +104,50 @@ func main() {
 	ctx := context.Background()
 	var g run.Group
 
-	var (
-		err error
-		wc  profilestorepb.ProfileStoreServiceClient = agent.NewNoopProfileStoreClient()
-		dc  debuginfo.Client                         = debuginfo.NewNoopClient()
-	)
+	var wc profilestorepb.ProfileStoreServiceClient = agent.NewNoopProfileStoreClient()
+	var dc debuginfo.Client = debuginfo.NewNoopClient()
 
 	if len(flags.StoreAddress) > 0 {
-		conn, err := grpcConn(reg, flags)
+		_, err := grpcConn(reg, flags)
 		if err != nil {
 			level.Error(logger).Log("err", err)
 			os.Exit(1)
 		}
-
-		wc = profilestorepb.NewProfileStoreServiceClient(conn)
-		dc = parcadebuginfo.NewDebugInfoClient(conn)
 	}
 
 	ksymCache := ksym.NewKsymCache(logger)
 
 	var (
-		pm            *agent.PodManager
-		sm            *agent.SystemdManager
-		targetSources = []agent.TargetSource{}
-		batcher       = agent.NewBatchWriteClient(logger, wc)
+		batcher  = agent.NewBatchWriteClient(logger, wc)
+		configs  discovery.Configs
+		listener = agent.NewProfileListener(logger, batcher)
 	)
 
 	if flags.Kubernetes {
-		pm, err = agent.NewPodManager(
-			logger,
-			flags.ExternalLabel,
-			node,
+		configs = append(configs, discovery.NewPodConfig(
 			flags.PodLabelSelector,
-			flags.SamplingRatio,
-			ksymCache,
-			batcher,
-			dc,
-			flags.TempDir,
 			flags.SocketPath,
-			flags.ProfilingDuration,
-		)
-		if err != nil {
-			level.Error(logger).Log("err", err)
-			os.Exit(1)
-		}
-		targetSources = append(targetSources, pm)
+			flags.Node,
+		))
 	}
 
 	if len(flags.SystemdUnits) > 0 {
-		sm = agent.NewSystemdManager(
-			logger,
-			node,
+		configs = append(configs, discovery.NewSystemdConfig(
 			flags.SystemdUnits,
-			flags.SamplingRatio,
-			flags.ExternalLabel,
-			ksymCache,
-			batcher,
-			dc,
-			flags.TempDir,
-			flags.ProfilingDuration,
-			flags.SystemdCgroupPath,
-		)
-		targetSources = append(targetSources, sm)
+		))
+	}
+
+	if flags.ExternalLabel == nil {
```

Comment:
How about we move this external label handling to a separate function? The main is already quite busy maybe we can try to refactor everything we touch a bit.